model/roles.go

```diff
@@ -373,38 +385,117 @@ func (m *RoleManifest) LookupRole(roleName string) *Role {
 func (m *RoleManifest) resolveLinks() validation.ErrorList {
 	errors := make(validation.ErrorList, 0)
 
-	// Resolve the consumers
+	// Build mappings of providers by name, and by type.  Note that the names
+	// involved here are the aliases, where appropriate.
+	providersByName := make(map[string]jobProvidesInfo)
+	providersByType := make(map[string][]jobProvidesInfo)
 	for _, role := range m.Roles {
-		// Prefer finding providers in the same role
-		providingRoles := append(Roles{role}, m.Roles...)
-		for _, job := range role.Jobs {
-		iterateOverConsumers:
-			for idx, consumer := range job.LinkConsumes {
-				for _, providingRole := range providingRoles {
-					for _, providingJob := range providingRole.Jobs {
-						for _, provider := range providingJob.LinkProvides {
-							if consumer.Type != "" && consumer.Type != provider.Type {
-								continue
-							}
-							if consumer.Name != "" && consumer.Name != provider.Name {
-								continue
-							}
-							if consumer.Name == "" {
-								consumer.Name = provider.Name
-							}
-							consumer.RoleName = providingRole.Name
-							consumer.JobName = providingJob.Name
-							continue iterateOverConsumers
-						}
-					}
+		for _, roleJob := range role.RoleJobs {
+			var availableProviders []string
+			for availableName, availableProvider := range roleJob.Job.AvailableProviders {
+				availableProviders = append(availableProviders, availableName)
+				providersByType[availableProvider.Type] = append(providersByType[availableProvider.Type], jobProvidesInfo{
```

Comment:
Throw an error if `availableProvider.Type` is the empty string?

Reply:
Actually, this should be done in `loadJobSpec`, both for `AvailableProviders` and `DesiredConsumers`.

Reply:
Hmm. Now I actually wonder if they _need_ to have types... I guess we'll enforce it for now and see what blows up.